features/192-asset-gear/index.md

```diff
@@ -0,0 +1,491 @@
+# Asset Gear
+
+## Overview
+
+The Asset Gear is a gear for uploading and retrieving user-generated asset.
+It supports both public and private asset.
+It can transform image asset on the fly via query string.
+
+## Layout
+
+The Asset Gear uses a single bucket for all apps. The asset name is in the following format.
+
+```
+/<app-id>/<random-string>[.<ext>]
+```
+
+- `app-id`: The app ID.
+- `random-string`: A randomly generated string.
+- `ext`: The inferred file extension (if any) from the given content type.
+
+## HTTP API
+
+### POST /_asset/presign
+
+#### Description
+
+This endpoint returns a presigned upload request which can be used to upload an asset.
+
+This endpoint requires API Key and authenticated user.
+
+#### Request JSON Schema
+
+```json
+{
+  "type": "object",
+  "properties": {
+    "content_type": { "type": "string" },
+    "content_md5": { "type": "string" },
+    "content_length": { "type": "integer" },
+    "access": { "enum": ["public", "private"] }
+  },
+  "required": ["content_type", "content_md5", "content_length"]
+}
+```
+
+- `content_type`: The media type of the asset.
+- `content_md5`: The MD5 hash of the asset.
+- `content_length`: The size in bytes of the asset.
+- `access`: The access control of the asset. `public` is the default.
```

Comment:
I’m fine with public and private at MVP, but can we have the spec (or at least refer to another feature issue from the spec) to support finer-grain permission control of users? (Since Skygear is suppose to tightly integrated with its auth)

Reply:
I have asked the team working on authorization but they are busy with other things and did not tell me what they have done.

If the access control of the asset has to take authentication into consideration, then accessing an asset could depend on the authentication status. Currently we have two session transport, header and cookie. When the developer writes a `<img>` with `src` set to an asset, the browser does not add the header. Therefore authenticated access to asset does not work with header session transport.

Reply:
I think the access control should not be in the asset url itself? For the asset, it only can be public or private. And the access control should be done by who can get the signed asset url. Currently master key is needed to get the signed url, so microservice need to wrap this api and control who will be enable to get the signed asset url. 

If we want asset to have finer-grain permission control, we should work on the `/_asset/sign` api instead but not the asset endpoint?